<a href="https://colab.research.google.com/github/poliane-poliane/Notas-de-aula-cientista-de-dados/blob/main/Bloco_1_Python_para_an%C3%A1lise_de_dados_%7C_M%C3%B3dulo_10_%7C_An%C3%A1lise_de_dados.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Módulo 10: Análise de dados


**[Bloco 1: Python para análise de dados]**

Vamos aprender a usar a Kaggle, uma plataforma em que se compartilha análises de dados.

Vamos passar pelas 3 etapas da nálise de dados. Vamos aprender a:



1.   **Explorar dados** (quais ferramentas se usa para ler e fazer análise inicial; aqui vamos usar o Pandas como um dos pacotes que se pode usar para fazer essa análise);
2.  **Transformação e limpeza de dados** (vamos aprender a normalizar o formato com que esse dados está armazenado nas colunas, vamos aprender a lidar com **dados faltantes**, garantir a limpeza dos dados para que estejam prontos
3.  **Visualização de dados** (vamos usar o Seaborn).

Por fim, vamos ver sobre **Storytelling**.

Na Kaggle, podemos compartilhar todos os nossos trabalhos de análise de dados e ir construindo um portfólio.

## Introdução à Kaggle

Kaggle é a maior comunidade online de ciência de dados e aprendizado de máquina. A plataforma permite que os usuários encontreme publiquem conjuntos de dados, construam e compartilhem notebooks (como este do Google Colab) e participem de competições e desafios de dados.

## Exploração de dados

Vamos explorar dados de crédito presentes [neste link](https://raw.githubusercontent.com/andre-marcos-perez/ebac-course-utils/develop/dataset/credito.csv).


Vamos explorar a partir de 3 verticals:



1.   Lógica do problema: qual assunto discutido com essa bas
2.   Estrutura do arquivo: texto, csv, quantidade de linhas e colunas
3.  Qualidade dos dados: base consistente, dados faltantes, etc.




Os dados estão no formato CSV e contém informações sobre clientes de uma instituição financeira. Em especial, estamos interessados em explicar a segunda coluna, chamada **default**, que indica se um cliente é adimplente(default = 0), ou inadimplente (default = 1), ou seja, queremos entender o porque um cliente deixa de honrar com suas dívidas baseado no comportamento de outros atributos, como salário, escolaridade e movimentação financeira. Uma descrição completa dos atributos está abaixo.

O atributo de interesse (default) é conhecido como variável resposta ou variável dependente, já os demais atributos que buscam explicá-la (idade, salário, etc.) são conhecidas como variáveis explicatívas, variáveis independentes ou até variáveis preditoras.


| Coluna  | Descrição |
| ------- | --------- |
| id      | Número da conta |
| default | Indica se o cliente é adimplente (0) ou inadimplente (1) |
| idade   | --- |
| sexo    | --- |
| depedentes | --- |
| escolaridade | --- |
| estado_civil | --- |
| salario_anual | Faixa do salario mensal multiplicado por 12 |
| tipo_cartao | Categoria do cartao: blue, silver, gold e platinium |
| meses_de_relacionamento | Quantidade de meses desde a abertura da conta |
| qtd_produtos | Quantidade de produtos contratados |
| iteracoes_12m | Quantidade de iteracoes com o cliente no último ano |
| meses_inatico_12m | Quantidade de meses que o cliente ficou inativo no último ano |
| limite_credito | Valor do limite do cartão de crédito |
| valor_transacoes_12m | Soma total do valor das transações no cartão de crédito no último ano |
| qtd_transacoes_12m | Quantidade total de transações no cartão de crédito no último ano |



Vamos começar lendo os dados num dataframe Pnadas.

In [ ]:
import pandas as pd

Com o pandas, é possível passar a base de dados só com o link de onde os dados estão. Affim, cria-se o data frame (df) mais facilmnte.

Há um parâmetro chamado `na_values` no `read_csv` que indica a representação de uma informação faltante (veremos com detalhes adiante)

In [ ]:
df = pd.read_csv('https://raw.githubusercontent.com/andre-marcos-perez/ebac-course-utils/develop/dataset/credito.csv', na_values = 'na')

In [ ]:
df.head(n=5)

id  default  idade sexo  dependentes         escolaridade  \
0  768805383        0     45    M            3         ensino medio   
1  818770008        0     49    F            5             mestrado   
2  713982108        0     51    M            3             mestrado   
3  769911858        0     40    F            4         ensino medio   
4  709106358        0     40    M            3  sem educacao formal   

  estado_civil   salario_anual tipo_cartao  meses_de_relacionamento  \
0       casado     $60K - $80K        blue                       39   
1     solteiro  menos que $40K        blue                       44   
2       casado    $80K - $120K        blue                       36   
3          NaN  menos que $40K        blue                       34   
4       casado     $60K - $80K        blue                       21   

   qtd_produtos  iteracoes_12m  meses_inativo_12m limite_credito  \
0             5              3                  1      12.691,51   
1             6              2                  1       8.256,96   
2             4              0                  1       3.418,56   
3             3              1                  4       3.313,03   
4             5              0                  1       4.716,22   

  valor_transacoes_12m  qtd_transacoes_12m  
0             1.144,90                  42  
1             1.291,45                  33  
2             1.887,72                  20  
3             1.171,56                  20  
4               816,08                  28

## Estrutura

In [ ]:
df.shape # retorna uma tupla (qtd linhas, qtd de colunas)

(10127, 16)

Vamos fazer um filtro para olhar dados sobre clientes adimplentes e inadimplentes.

In [ ]:
# esse é um filtro booleano visto no módulo 7
# do fata frame origigal df[...] exiba as informações de quando as colunas do df default df['default'] é igual a 0
# exiba o shape (num de linha e colunas)

df[df['default'] == 0].shape

(8500, 16)

In [ ]:
df[df['default'] == 1].shape

(1627, 16)

Em vez de pedir número de linhas e colunas, você pode pedir para ver o cabeçalho desse novo dataframe (`df[df['default'] == 0]` ou `df[df['default'] == 1]`) criado para essa condição particular:

In [ ]:
df[df['default'] == 1].head(n=4)

id  default  idade sexo  dependentes escolaridade estado_civil  \
21  708508758        1     62    F            0     mestrado       casado   
39  708300483        1     66    F            0    doutorado       casado   
51  779471883        1     54    F            1     mestrado       casado   
54  714374133        1     56    M            2     mestrado       casado   

     salario_anual tipo_cartao  meses_de_relacionamento  qtd_produtos  \
21  menos que $40K        blue                       49             2   
39             NaN        blue                       56             5   
51  menos que $40K        blue                       40             2   
54         $120K +        blue                       36             1   

    iteracoes_12m  meses_inativo_12m limite_credito valor_transacoes_12m  \
21              3                  3       1.438,73               692,25   
39              3                  4       7.882,47               704,78   
51              1                  3       1.438,08               705,59   
54              3                  3      15.769,34               602,60   

    qtd_transacoes_12m  
21                  16  
39                  16  
51                  19  
54                  15

In [ ]:
qtd_total, _ = df.shape
qtd_adimpentes, _ = df[df['default'] == 0].shape
qtd_inadimpentes, _ = df[df['default'] == 1].shape
#_, qtd_inadimpentes = df[df['default'] == 1].shape

print(f'Quantidade total de clientes: {qtd_total}')
print(f'Quantidade de clientes adimplentes: {qtd_adimpentes}')
print(f'Quantidade de clientes inadimplentes: {qtd_inadimpentes}\n')

print(f'A proporção de clientes adimplentes é de {round(100 * qtd_adimpentes / qtd_total, 2)}%')
print(f'A proporção de clientes inadimplentes é de {round(100 * qtd_inadimpentes / qtd_total, 2)}%')

Quantidade total de clientes: 10127
Quantidade de clientes adimplentes: 8500
Quantidade de clientes inadimplentes: 1627

A proporção de clientes adimplentes é de 83.93%
A proporção de clientes inadimplentes é de 16.07%


Uma coisa interessante e **nova** é o undeline em `qtd_total, _ `.

**Esse caractere ignora** o número de colunas (já que ele é desecessário aqui).

Entretanto o uso do `underline`

## Schema (Qualidade dos dados)

Aqui vamos relacionar o tipo da informação de uma coluna com o tipo de dado que está nela. Então, por exemplo, no caso de default, as variáveis são 0 e 1, indicando se a pessoa é adimplente ou inadimplente. Portanto, para representa essas duas informações, faz sentido haver só duas possibilidades mesmo: 0 ou 1 (strings ou variáveis booleanas). Poderia se A e I (caracteres). O que não faria sentido era haver um terceiro tipo de dado, seja uma letra , seja um número, porque isso não estaria de acordo com o tipo de informações que eu preciso naquela coluna. Um outro exemplo, a coluna sexo: se o conceito de sexo foi partir de apenas duas possibilidades - feminino e masculino - só faz sentido ter F e M ou 0 e 1; se houver outro dado, pode ser que haja inconsistência. Outro exemplo, a coluna de idade: só faz sentido haver inteiros. E assim por diante. É preciso manter a consistência.

In [ ]:
df.head(n=5)

id  default  idade sexo  dependentes         escolaridade  \
0  768805383        0     45    M            3         ensino medio   
1  818770008        0     49    F            5             mestrado   
2  713982108        0     51    M            3             mestrado   
3  769911858        0     40    F            4         ensino medio   
4  709106358        0     40    M            3  sem educacao formal   

  estado_civil   salario_anual tipo_cartao  meses_de_relacionamento  \
0       casado     $60K - $80K        blue                       39   
1     solteiro  menos que $40K        blue                       44   
2       casado    $80K - $120K        blue                       36   
3          NaN  menos que $40K        blue                       34   
4       casado     $60K - $80K        blue                       21   

   qtd_produtos  iteracoes_12m  meses_inativo_12m limite_credito  \
0             5              3                  1      12.691,51   
1             6              2                  1       8.256,96   
2             4              0                  1       3.418,56   
3             3              1                  4       3.313,03   
4             5              0                  1       4.716,22   

  valor_transacoes_12m  qtd_transacoes_12m  
0             1.144,90                  42  
1             1.291,45                  33  
2             1.887,72                  20  
3             1.171,56                  20  
4               816,08                  28

Para listar o tipo de dado de cada coluna, usamos o `dtypes`:

In [ ]:
df.dtypes

id                          int64
default                     int64
idade                       int64
sexo                       object
dependentes                 int64
escolaridade               object
estado_civil               object
salario_anual              object
tipo_cartao                object
meses_de_relacionamento     int64
qtd_produtos                int64
iteracoes_12m               int64
meses_inativo_12m           int64
limite_credito             object
valor_transacoes_12m       object
qtd_transacoes_12m          int64
dtype: object

`object`é uma string.

Após rodar o `dtypes`, é importante olhar para o resultado do print e checar se estão todos de acordo com o esperado.

Veja que as variáveis `limite_credito`e `valor_transacoes_12m`estão como `object`, o que é inesperado, pois é esperado que fossem `float`. Ao olhar para os dados da tabela (`df.head(n=5)`), notamos que isso aconteceu porque os dados estão escritos no formato brasileiro de escrever números: `NN.NNN,NN`; **precisamos remover** o ponto e colocar ponto no lugar da vírgula.

### Método `describe()` e atributos categóricos e numéricos

O método `describe()` é um método do pandas que permite fazer resumos de dados e exibir de forma visual.

Vamos ver abaixo como usá-lo aplicando para variáveis tipo `string` e tipo `number`.

Para isso, vamos usar o `select_dtypes` para que esse resumo dos dados seja feito com relação às colunas que possuem o tipo de dado selecionado. O `transpose`vai ser usado para melhor visualização dos dados.

O método `describe()`nos ajuda a avaliar a consistência dos dados, no sentido de detalhar o tipo de dados em cada coluna e de nos permitir avaliar se os dados estão completos ou faltantes, além de permitir fazer uma visualização analítica prévia dos dados: saber quais são os maiores tipos, etc.


O `describe()` exibe colunas com atributos difrentes para variáveis categóricas e para variáveis numéricas.

Para **variáveis categóricas**, o `describe()` exibe colunas com as informações:




1.   **count**

2.   **unique**

3.    **top**

4.  **freq**


Para **variáveis categóricas**, o `describe()` exibe colunas com as informações:




1.   **count**

2.   **mean** (média)

3.    **std** (standard deviation, ou desvio padrão)

4.  **min** e **max** (mínimo e máximo)

5.  **25%**, **50%**, **75%** (quartis)




*   **Atributos categóricos**

São aqueles cujo conjunto de valores é finito. Exemplo: a variável sexo só pode ter 2 valores: F ou M.

Vamos ver o `describe`aplicado a dados tipo `object`, sem o `transpose`, por enquanto:





In [ ]:
df.select_dtypes('object').describe()

sexo escolaridade estado_civil   salario_anual tipo_cartao  \
count   10127         8608         9378            9015       10127   
unique      2            5            3               5           4   
top         F     mestrado       casado  menos que $40K        blue   
freq     5358         3128         4687            3561        9436   

       limite_credito valor_transacoes_12m  
count           10127                10127  
unique           9272                10035  
top          1.438,21             3.851,51  
freq               11                    3


Vamos ver o `describe`aplicado a dados tipo `object` com o `transpose`. Dessa forma, fica bem melhor de avaliar os dados.

In [ ]:
df.select_dtypes('object').describe().transpose()

count unique             top  freq
sexo                  10127      2               F  5358
escolaridade           8608      5        mestrado  3128
estado_civil           9378      3          casado  4687
salario_anual          9015      5  menos que $40K  3561
tipo_cartao           10127      4            blue  9436
limite_credito        10127   9272        1.438,21    11
valor_transacoes_12m  10127  10035        3.851,51     3

Para **variáveis categóricas**, o `describe()` exibe colunas com as informações:




1.   **count:** faz a contagem dos dados para uma dada variável. Pode ajudar a verificar que a quantidade de dados está de acordo com o número de linhas do dataframe. Se houver menos dados que o número de linha, pode sugerir dados faltantes ou de outra natureza.

2.   **unique:** indica quantos tipos diferentes te. Se o tipo de dado for booleano, indica só 2; se o tipo de dado for, por exemplo, letras A, B, C, D, E, irá indicar 5; se o tipo de dado for float, vai contar o número de valores únicos.

3.    **top:** indica qual, dentre os dados disponíveis para aquela coluna, é o dado que mais se repete. Se entre A, B, C, D, E, o dado que mais se repete entre os dados de uma dada coluna é a letra A, então ele será o valor que vai aparecer em top para aquela coluna.

4.  **freq:** indica a frequência com que o dado em top se repete. Então se A é o dado que mais se repete em uma coluna, o dado de freq indica quantas vezes a letra A se repete.


Vejamos que:



*   **count:** para a coluna sexo está de acordo com o número de linhas do `df`, como vimos ao rodar o `df.shape`: 10127 linhas e 16 colunas. Para as colunas escolaridade, estado_civil e salario_anual, esse valor é menor que 10127, indicando que pode haver dados faltantes.

*   **unique:** para a coluna sexo, há dois valores, indicando que faz sentido, já que estamos com as possibilidade F e M para essa coluna. As demais também estão ok.

*  **top:** na coluna sexo, o valor que mais aparede é F. Na coluna escolaridade, mestrado. Na coluna salario_anual, a maior parte é de pessoas que recebem menos de 40K. E assim por diante.

*   **freq:** na coluna sexo, F aparece 5358 vezes. E assim por diante. Poderíamos até dividir 5358 por 10127 para vermos quantos por cento F aparece entre os sexos. Veja que a coluna salario_anual é formado pela maior parte de pessoas com menos de 40K.



*   **Atributos numéricos**

São aqueles cujo conjunto de valores podem assumir infinitos valores. Exemplo: as variáveis tipo inteiro e float.

Vamos escrever o mesmo para as variáveis numéricas:

In [ ]:
df.select_dtypes('number').describe().transpose()

count          mean           std          min  \
id                       10127.0  7.391776e+08  3.690378e+07  708082083.0   
default                  10127.0  1.606596e-01  3.672348e-01          0.0   
idade                    10127.0  4.632596e+01  8.016814e+00         26.0   
dependentes              10127.0  2.346203e+00  1.298908e+00          0.0   
meses_de_relacionamento  10127.0  3.592841e+01  7.986416e+00         13.0   
qtd_produtos             10127.0  3.812580e+00  1.554408e+00          1.0   
iteracoes_12m            10127.0  2.455317e+00  1.106225e+00          0.0   
meses_inativo_12m        10127.0  2.341167e+00  1.010622e+00          0.0   
qtd_transacoes_12m       10127.0  6.485869e+01  2.347257e+01         10.0   

                                 25%          50%          75%          max  
id                       713036770.5  717926358.0  773143533.0  828343083.0  
default                          0.0          0.0          0.0          1.0  
idade                           41.0         46.0         52.0         73.0  
dependentes                      1.0          2.0          3.0          5.0  
meses_de_relacionamento         31.0         36.0         40.0         56.0  
qtd_produtos                     3.0          4.0          5.0          6.0  
iteracoes_12m                    2.0          2.0          3.0          6.0  
meses_inativo_12m                2.0          2.0          3.0          6.0  
qtd_transacoes_12m              45.0         67.0         81.0        139.0

Veja que o `id` está configurado como inteiro, mas não faz sentido, já que é só um numero de identificação, poderíamos muito bem tratá-lo como string, mas sua configuração de dados veio como `int`; por isso, no `describe()`, seus valores são imensos. Para ignorar a coluna de ids, podemos usar o método

```
# drop()
```

 como no código abaixo.

|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||


|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||

|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||


#### Método `drop()`

 O `drop()` é usado para remover linhas ou colunas de um dataframe. Sua anatomia geral é:



```
# dataframe.drop(labels, axis = 0, inplace = False)
```



*   **labels**: Nome da coluna(s) ou índice(s) que você quer remover.
*   **axis**: Define se está removendo uma linha (`axis = 0`) ou uma coluna (`axis = 1`)
*   **inplace**: Se `True`, altera o dataframe origina. Se `False` (padrão), retorna uma cópia sem os elementos removidos.


In [ ]:
df.drop('id', axis = 1).select_dtypes('number').describe().transpose()

count       mean        std   min   25%   50%  \
default                  10127.0   0.160660   0.367235   0.0   0.0   0.0   
idade                    10127.0  46.325960   8.016814  26.0  41.0  46.0   
dependentes              10127.0   2.346203   1.298908   0.0   1.0   2.0   
meses_de_relacionamento  10127.0  35.928409   7.986416  13.0  31.0  36.0   
qtd_produtos             10127.0   3.812580   1.554408   1.0   3.0   4.0   
iteracoes_12m            10127.0   2.455317   1.106225   0.0   2.0   2.0   
meses_inativo_12m        10127.0   2.341167   1.010622   0.0   2.0   2.0   
qtd_transacoes_12m       10127.0  64.858695  23.472570  10.0  45.0  67.0   

                          75%    max  
default                   0.0    1.0  
idade                    52.0   73.0  
dependentes               3.0    5.0  
meses_de_relacionamento  40.0   56.0  
qtd_produtos              5.0    6.0  
iteracoes_12m             3.0    6.0  
meses_inativo_12m         3.0    6.0  
qtd_transacoes_12m       81.0  139.0

Para **variáveis categóricas**, o `describe()` exibe colunas com as informações:




1.   **count**

2.   **mean** (média)

3.    **std** (standard deviation, ou desvio padrão)

4.  **min** e **max** (mínimo e máximo)

5.  **25%**, **50%**, **75%** (quartis)

Agora, com esses dados, podemos fazer uma análise prévia sobre os dados das colunas com valores numéricos.


*   Como o count de todas as colunas selecionadas é igual ao número de linhas do datafram (como vimos do `df.shape), então aparentemente não temos problemas de consistência com os dados dessas colunas.
*   Veja que a coluna `default`está bem consistente, já que os dados nela só podem ser o ou 1. Para essa coluna, a média (`mean`) é de 0.16, indicando que 16% são de inadimplentes (adimplente recebe valor 1). **Isto porque a média de uma coluna binária é igual à proporção dos 1's**.

***Veja com mais detalhes na seção Curiosidades, a úntima seção desse módulo.***


Mas não vamos entrar em muitos detalhes dessa análise, até porque estamos na etapa de tratar os dados.



|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||


|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||

|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||


## Dados faltantes

Dados faltantes podem ser:



*   Vazios ou string vazia ("");
*   Nulos (`None`);
*   Não disponíveis ou aplicáveis (`na`, `NA`, etc.);
*   Não numérico (`nan`, `NAN`, `NaN`, etc.).

Em princípio, é muito difícil trabalhar com dados faltantes, porque não sabemos como ele ve.

Inclusive, quando definimos o dataframe no início desse módulo como



```
# df = pd.read_csv('https://raw.githubusercontent.com/andre-marcos-perez/ebac-course-utils/develop/dataset/credito.csv', na_values = 'na')
```

usamod o parâmetro `na_values`, próprio da função `pd.read_csv()`para indicar quais valores no arquivo CSV devem ser interpretados como **valores ausentes (NA, Not available, nulos, etc.).**


### Método `isna()`e `any()`

Podemos verificar quais colunas possuem dados faltantes usando o método `isna()` para localizar os dados faltantes e o método àny()`para filtrar as colunas de interesse:

In [ ]:
df.isna().any()

id                         False
default                    False
idade                      False
sexo                       False
dependentes                False
escolaridade                True
estado_civil                True
salario_anual               True
tipo_cartao                False
meses_de_relacionamento    False
qtd_produtos               False
iteracoes_12m              False
meses_inativo_12m          False
limite_credito             False
valor_transacoes_12m       False
qtd_transacoes_12m         False
dtype: bool

As colunas`escolaridade`, `estado_civil`e `salario_anual` acusaram `True`, indicando que elas possuem dados faltantes.

Mas **como descobrir a quantidade de daods faltantes**?

Vamos definir a seguinte função que irá imprimir qual é essa quantidade de dados faltantes nas 3 colunas que acusaram `True` e a porcentagem.

**Note que essa função retorna `None`porque ela retorna um print, então não é para retornar nada mesmo.**

In [ ]:
df.columns

Index(['id', 'default', 'idade', 'sexo', 'dependentes', 'escolaridade',
       'estado_civil', 'salario_anual', 'tipo_cartao',
       'meses_de_relacionamento', 'qtd_produtos', 'iteracoes_12m',
       'meses_inativo_12m', 'limite_credito', 'valor_transacoes_12m',
       'qtd_transacoes_12m'],
      dtype='object')

ou ainda, sem o `Index`:

In [ ]:
df.columns.to_list()

['id',
 'default',
 'idade',
 'sexo',
 'dependentes',
 'escolaridade',
 'estado_civil',
 'salario_anual',
 'tipo_cartao',
 'meses_de_relacionamento',
 'qtd_produtos',
 'iteracoes_12m',
 'meses_inativo_12m',
 'limite_credito',
 'valor_transacoes_12m',
 'qtd_transacoes_12m']

Agora, vamos fazer uma espécie de loop columa por coluna na função abaixo para achamaos os valores faltantes:

In [ ]:
def stats_dados_faltantes(df: pd.DataFrame) -> None:

  stats_dados_faltantes = []
  for col in df.columns:
    if df[col].isna().any():
      qtd, _ = df[df[col].isna()].shape
      total, _ = df.shape
      dict_dados_faltantes = {col: {'quantidade': qtd, 'porcentagem': round(100 * qtd / total, 2)}}
      stats_dados_faltantes.append(dict_dados_faltantes)

  for stat in stats_dados_faltantes:
    print(stat)

Chamando a função para operar sobre o nosso df:

In [ ]:
stats_dados_faltantes(df = df)

{'escolaridade': {'quantidade': 1519, 'porcentagem': 15.0}}
{'estado_civil': {'quantidade': 749, 'porcentagem': 7.4}}
{'salario_anual': {'quantidade': 1112, 'porcentagem': 10.98}}


Ou seja:



*   Em **escolaridade** temos 1519 dados faltantes, ou seja, 15% da nossa base na coluna de escolaridade
*   Em **estado_civil** temos 7.4% de dados faltantes na base dessa coluna.
*   Em **salario_anual** temos 10.98% de dados faltantes na base dessa coluna.


Como nossa base está desbalanceada (inadimplentes e adimplentes), vamos ver como essa falta de dados se dá para cada tipo de cliente:


In [ ]:
stats_dados_faltantes(df = df[df['default'] == 0])

{'escolaridade': {'quantidade': 1263, 'porcentagem': 14.86}}
{'estado_civil': {'quantidade': 620, 'porcentagem': 7.29}}
{'salario_anual': {'quantidade': 925, 'porcentagem': 10.88}}


In [ ]:
stats_dados_faltantes(df = df[df['default'] == 1])

{'escolaridade': {'quantidade': 256, 'porcentagem': 15.73}}
{'estado_civil': {'quantidade': 129, 'porcentagem': 7.93}}
{'salario_anual': {'quantidade': 187, 'porcentagem': 11.49}}


Por que queremos verificar isso?

Note que o que vimos nas 2 linhas de código acima foi que o percentual de dados faltantes tanto para adimplentes quanto para inadimplentes é bem próximo, ou seja, é bem equilibrado: para a coluna de escolaridade, 14.86% dos dados faltantes são de de inadimplentes (0) e 15.73% são de adimplentes (1), ou seja, são bem próximos; para estado_civil e salario_anual a dfierença também é pequena.

Quando essa diferença é pequena, podemos tratar esses dados faltantes **por meio da remoção dessas linhas com dados faltantes**. Sim, vamos perder: cerca de 15% dos dados da base da coluna escolaridade; cerca de 7% dos dados da base da coluna estado_civil e cerca de 11% dos dados da base da coluna salario_anual.

**Mas isso é delicado!!!!!!!!!!!!!!!!!!!!!!**

Caso a maior parte dos dados de um ou outro, isso seria bem ruim e o tratamento desses dados faltantes fosse pela via da remoção das linhas de dados faltantes seria enviesado.

|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||


|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||

|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||

## Tranformação e limpeza de dados

Agoraque conhecemos melhor a natureza do nosso conjunto de dados, vamos conduzir uma atividade conhecida como ***data wrangling***, que consiste na transformação e limpeza dos dados do conjunto para que possam ser melhor analisados. Em especial, vamos remover:



*   Corrigir o *schema* das nossas colunas (converter os dados das colunas que não tipo strig no que deveriam ser: float);
*   Remover os dados faltantes.




### Correção do schema

Na etapa de exploração, notamos que as colunas **limite_credito** e **valor_transacoes_12m** estavam sendo interpretadas como colunas categóricas (`dtype = object`).

As duas últimas linhas de código abaixo são apenas para revermos os `dtypes`de cada uma dessas colunas e dar uma olhada no cabeçalho do dataframe apenas para essas colunas.

In [ ]:
df.dtypes

id                          int64
default                     int64
idade                       int64
sexo                       object
dependentes                 int64
escolaridade               object
estado_civil               object
salario_anual              object
tipo_cartao                object
meses_de_relacionamento     int64
qtd_produtos                int64
iteracoes_12m               int64
meses_inativo_12m           int64
limite_credito             object
valor_transacoes_12m       object
qtd_transacoes_12m          int64
dtype: object

In [ ]:
# Para vermos o dtypes só das colunas selecionadas, basta usar o df.dtypes só que agora passando a lista com as colunas desejadas
df[ ['limite_credito', 'valor_transacoes_12m'] ].dtypes

limite_credito          object
valor_transacoes_12m    object
dtype: object

In [ ]:
df[ ['limite_credito', 'valor_transacoes_12m'] ].head(n=5)

limite_credito valor_transacoes_12m
0      12.691,51             1.144,90
1       8.256,96             1.291,45
2       3.418,56             1.887,72
3       3.313,03             1.171,56
4       4.716,22               816,08

Veja que, de fato, os dados acima estão com o formato brasileiro de números. O que vamos fazer é remover os pontos e trocar as vírgulas por ponto.

Para isso, vamos criar uma função `lambda`para limpar os dados.

Mas antes, vamos **testar** como ela se aplica aos dados apenas da coluna limite_credito através do método funcional `map`.

Só relembrando que:




*   Anatomia da função `lambda`


```
 variavel = lambda parametros: expressao
```


*   Anatomia da função `map`


```
 variavel = map(funcao , colecao)
```

Então vamos criar uma lista com os dados dessa coluna e aplicar a função `map`. Se virmos que o teste deu certo, aí, sim, vamos aplicar ao nosso dataframe.

Primeiro, vamos criar a lista com alguns valores da coluna limite_credito; podemos considerar os 5 primeiros valores.

In [ ]:
valores_originais = df['limite_credito'].head(n=5)
valores_originais = list(valores_originais)
print(valores_originais)

['12.691,51', '8.256,96', '3.418,56', '3.313,03', '4.716,22']


O print foi só para vermos que dessa forma é possível selecionar só alguns valores da coluna, em vez de todos os valores. Caso quiséssemos todos, bastava remover o `head`.

Agora vamos escrever a função `lambda`.

Queremos

*   remover os pontos, que é *equivalente a* **substituir** ponto por nada/vazio; e
*   **substituir** vírgula por ponto.


Para isso, vamos usar a função `replace()` **DUAS VEZES**.


In [ ]:
funcao_teste = lambda valor: float(valor.replace(".","").replace(",","."))

# Chamando a função através da função map
valores_limpos = map(funcao_teste, valores_originais)
print(list(valores_limpos))

[12691.51, 8256.96, 3418.56, 3313.03, 4716.22]


Portanto vimos que **FUNCIONOU!!!!**

Vamos compilar tudo em uma célula de código:

In [ ]:
# 1º: listar a coleção sobre a qual vamos aplicar a função de subtituição
valores_originais = list(df['limite_credito'].head(n=5))

# 2º: definindo a função de substituição
funcao = lambda valor: float( valor.replace(".","").replace(",",".") )

# 3º: aplicando a função de substituição à coleção a ser transformada e convertendo em lista
valores_limpos = list(map(funcao, valores_originais))

print(valores_limpos)

[12691.51, 8256.96, 3418.56, 3313.03, 4716.22]


Agora que vimos que funcionou, vamos aplicar isso aos dados das duas colunas com problema, usando a **função** `apply()`.

In [ ]:
df['valor_transacoes_12m'] = df['valor_transacoes_12m'].apply(funcao)
#df['valor_transacoes_12m'].head(n=8)

In [ ]:
df['limite_credito'] = df['limite_credito'].apply(funcao)

Vamos ver o schema novamente para ver que agora são tipo float:

In [ ]:
df.dtypes

id                           int64
default                      int64
idade                        int64
sexo                        object
dependentes                  int64
escolaridade                object
estado_civil                object
salario_anual               object
tipo_cartao                 object
meses_de_relacionamento      int64
qtd_produtos                 int64
iteracoes_12m                int64
meses_inativo_12m            int64
limite_credito             float64
valor_transacoes_12m       float64
qtd_transacoes_12m           int64
dtype: object

Abora vamos rodar o `describe()`para vermos se a consistência dos dados foi mantida em termos da comparação entre número de dados e linhas:

In [ ]:
df.select_dtypes('object').describe().transpose()

count unique             top  freq
sexo           10127      2               F  5358
escolaridade    8608      5        mestrado  3128
estado_civil    9378      3          casado  4687
salario_anual   9015      5  menos que $40K  3561
tipo_cartao    10127      4            blue  9436

In [ ]:
df.drop('id', axis = 1).select_dtypes('number').describe().transpose()

count         mean          std      min       25%  \
default                  10127.0     0.160660     0.367235     0.00     0.000   
idade                    10127.0    46.325960     8.016814    26.00    41.000   
dependentes              10127.0     2.346203     1.298908     0.00     1.000   
meses_de_relacionamento  10127.0    35.928409     7.986416    13.00    31.000   
qtd_produtos             10127.0     3.812580     1.554408     1.00     3.000   
iteracoes_12m            10127.0     2.455317     1.106225     0.00     2.000   
meses_inativo_12m        10127.0     2.341167     1.010622     0.00     2.000   
limite_credito           10127.0  8632.440165  9088.788041  1438.00  2555.355   
valor_transacoes_12m     10127.0  4404.583047  3397.128078   510.16  2155.825   
qtd_transacoes_12m       10127.0    64.858695    23.472570    10.00    45.000   

                             50%        75%       max  
default                     0.00      0.000      1.00  
idade                      46.00     52.000     73.00  
dependentes                 2.00      3.000      5.00  
meses_de_relacionamento    36.00     40.000     56.00  
qtd_produtos                4.00      5.000      6.00  
iteracoes_12m               2.00      3.000      6.00  
meses_inativo_12m           2.00      3.000      6.00  
limite_credito           4549.42  11068.175  34516.99  
valor_transacoes_12m     3899.59   4741.310  18484.93  
qtd_transacoes_12m         67.00     81.000    139.00

Encerrado isso, note que abaixo anotei uma **OUTRA FORMA DE** aplicar a função de substituição (map) nas colunas para converter os dados. **PORÉM** essa forma não é tão eficiente quanto o `apply()`. Não rodei porque vai acusar erro, já que os dados já foram convertidos para float.

In [ ]:
transacoes_12m_limpas = list(map(funcao, list(df['valor_transacoes_12m']) ))
print(transacoes_12m_limpas)

|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||


|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||

|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||

### Remoção dos dados faltantes

Já ao ler o conjunto de fados, indicamos ao Python uma forma de interpretar dados faltantes por meio de



```
na_values = 'na'
```

de modo que, toda vez que ele encontra um dado faltante ele substitui por `NaN`.

Sendo assim, como o pandas, portanto, está ciente de um dado faltante, a remoção das linhas problemáticas é trivial.

Para isso, usamos o método `dropna`, para ignirar/pular as linhas com dados faltantes.

Sobre esse `df.dropna(inplace=True)`, vale algumas informações:



*   ele altera o próprio dataframe **diretamente**, sem criar uma cópia.
*   ou seja, os valores ausentes  são removidos do próprio df, e ele é modificado em memória.
*   não retorna nada (retorna `None`), pois a operação é feita "in place" (no lugar).



In [ ]:
df.dropna(inplace=True)

Vamos analisar a estrutura dos dados novamente agora que removemos os dados ausentes:

In [ ]:
df.shape

(7081, 16)

Veja que saímos de 10127 linhas para 7081 linhas.

In [ ]:
# Adimplentes
df[ df['default'] == 0 ].shape

(5968, 16)

In [ ]:
# Inadimplentes
df[ df['default'] == 1 ].shape

(1113, 16)

Veja que essa perda de linhas também teve impacto nos adimplentes e inadimplentes.

Vamos ver isso em termos percentuais.

In [ ]:
qtd_total_novo = df.shape[0] #[0] indica nº de linhas do df
qtd_adimplentes_novo = df[ df['default'] == 0].shape[0]
qtd_inadimplentes_novo = df[ df['default'] == 1].shape[0]

Outra forma de obtermos os dados logo acima seria usando o **underline**:

In [ ]:
qtd_tot_novo, _ = df.shape
qtd_adimp_novo, _ = df[ df['default'] == 0 ].shape
qtd_inadimp_novo, _ = df[ df ['default'] == 1 ].shape

Essas 2 formas são:

**Funcionalmente, são exatamente equivalentes**. *A diferença é puramente de estilo e legibilidade. Ambas estão corretas*.

Calculando as porcentagens:

In [ ]:
perc_adimp_novo = round((qtd_adimp_novo / qtd_tot_novo)*100,2)
perc_inadimp_novo = round((qtd_inadimp_novo / qtd_tot_novo)*100,2)

print(f'Nova Proporção de Adimplentes: {perc_adimp_novo}%')
print(f'Nova proporção de Inadimplentes: {perc_inadimp_novo}%')

Nova Proporção de Adimplentes: 84.28%
Nova proporção de Inadimplentes: 15.72%


|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||


|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||

|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||


### 🚨Diferença entre `df.dropna` e `df.dropna(inplace=True)`


Vimos que para remover as linhas faltantes utilizamos o



```
df.dropna(inplace=True) ## remove NaNs direto no df; df agora não tem mais linhas com NaN
```


Porém o professor também como a possibilidade de usar a forma

```
df = df.dropna() ## # cria um novo dataframe sem NaNs e salva na variável df; df agora é o novo dataframe sem NaNs
```
mas que dessa forma estaríamos reatribuindo o novo df para df.

Em consulta ao ChatGPT, vi que esses 2 formatos têm as seguintes diferenças:


*   `df.dropna(inplace=True)`

1.   Altera o próprio dataframe df diretamente, sem criar uma cópia.
2.   Ou seja, os valores ausentes (NaN) são removidos no próprio objeto df, e ele é modificado em memória.
3.   Não retorna nada (retorna None), pois a operação é feita "in place" (no lugar).

*   `df = df.dropna()`

1.   Cria um novo dataframe, que é o resultado de remover as linhas com NaN, e atribui esse novo dataframe à variável df.
2.   O dataframe antigo (antes de remover os NaNs) é sobrescrito por esse novo, mas o processo ocorre via atribuição, não via alteração direta.

#### Qual usar?


🟩 Se você quer alterar o dataframe diretamente, sem precisar criar outro, use inplace=True.

🟦 Se quer um código mais claro, seguro, e evitar efeitos colaterais não intencionais, prefira df = df.dropna().

🔧 A própria documentação do pandas recomenda evitar inplace=True, pois em muitas situações a atribuição direta (df = df.dropna()) é mais clara, previsível e até mais eficiente internamente.



| Forma                     | Efeito no dataframe | Retorna valor? | Estilo                  |
| ------------------------- | ------------------- | -------------- | ----------------------- |
| `df.dropna(inplace=True)` | Altera diretamente  | ❌ None         | Efeito colateral direto |
| `df = df.dropna()`        | Cria uma nova cópia | ✅ DataFrame    | Mais explícito e seguro |


|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||


|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||

|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||


### 🚨 Caso em que ao menos um dos dados seja float

No caso em que estudamos acima, as 2 colunas com problema tinha valores tipo object (str em pandas).

Mas se algumas entradas fossem tipo float, **NÃO CONSEGUIRÍAMOS aplicar operações como `replace` ou conversões como `map`**. Para contornar isso, precisaríamos tratar cada caso previamente, como segue:

In [ ]:
def tratar_valor(x):
    if isinstance(x, float) or isinstance(x, int):
        return x
    if isinstance(x, str):
        x = x.replace("R$", "").replace(".", "").replace(",", ".")
        return float(x)
    return None  # ou outra forma de tratar valores inesperados

df['limite_credito'] = df['limite_credito'].map(tratar_valor)
df['valor_transacoes_12m'] = df['valor_transacoes_12m'].map(tratar_valor)

Assim, a função map funcionará corretamente independentemente do tipo dos valores na coluna.

|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||


|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||

|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||



# Curiosidades


## Método drop e variáveis numéricas

Ao utilizar o código

In [ ]:
df.drop('id', axis = 1).select_dtypes('number').describe().transpose()

count       mean        std   min   25%   50%  \
default                  10127.0   0.160660   0.367235   0.0   0.0   0.0   
idade                    10127.0  46.325960   8.016814  26.0  41.0  46.0   
dependentes              10127.0   2.346203   1.298908   0.0   1.0   2.0   
meses_de_relacionamento  10127.0  35.928409   7.986416  13.0  31.0  36.0   
qtd_produtos             10127.0   3.812580   1.554408   1.0   3.0   4.0   
iteracoes_12m            10127.0   2.455317   1.106225   0.0   2.0   2.0   
meses_inativo_12m        10127.0   2.341167   1.010622   0.0   2.0   2.0   
qtd_transacoes_12m       10127.0  64.858695  23.472570  10.0  45.0  67.0   

                          75%    max  
default                   0.0    1.0  
idade                    52.0   73.0  
dependentes               3.0    5.0  
meses_de_relacionamento  40.0   56.0  
qtd_produtos              5.0    6.0  
iteracoes_12m             3.0    6.0  
meses_inativo_12m         3.0    6.0  
qtd_transacoes_12m       81.0  139.0

na seção *Método drop()*, notei que a média (mean) para a coluna de dados default estava indicando 0.16, registrando que, entre os adimplentes (0) e inadimplentes(1), os adimplentes correspondem a 16%. Isto porque, no caso de uma coluna binária (0 e 1), **a média é exatamente igual à proporção de 1's**.

Portanto, o cálculo seria assim:

In [ ]:
proporcao_inadimplentes = df['default'].mean()
print(proporcao_inadimplentes)

0.1606596227905599


### Value_counts() em termos percentuais

Caso queiramos ver a proporção de cada um (0 ou 1), podemos usar o `value_counts`:

In [ ]:
print(df['default'].value_counts(normalize=True))

default
0    0.83934
1    0.16066
Name: proportion, dtype: float64


### Value_counts() como contagem de dados

Caso queiramos ver a quantidade de dados de cada um (0 ou 1), podemos usar o `value_counts` sem parâmetro:

In [ ]:
print(df['default'].value_counts())

default
0    8500
1    1627
Name: count, dtype: int64


## UNDERLINE ()

O `underline` (_) é uma convenção usada para indicar que você **NÃO VAI USAR** aquele valor. Ou seja, em

In [ ]:
qtd_total_pte1, _ = df.shape
print(qtd_total_pte1)

10127


significa que:

***Quero guardar o primeiro valor (número de linhas) em qtd_total, e ignorar o segundo valor (número de colunas).***

Escrevendo ao contrário, vamos guardar o número de colunas:

In [ ]:
_ , qtd_total_pte2 = df.shape
print(qtd_total_pte2)

16


### Outros exemplos de usos do underline()



*   Quando você quer o segundo valor de uma tupla/lista


In [ ]:
valores = ['a', 'b', 'c']

for _, letra in enumerate(valores):
    print(letra)

a
b
c



*   Quando uma função retorna dois valores e você só quer o segundo

In [ ]:
def retorna_dois():
    return 10, 20

_, segundo = retorna_dois()
print(segundo)  # → 20

20



*   Ao iterar com desempacotamento e só precisa do segundo valor

In [ ]:
dicionario = {'a': 1, 'b': 2}
for _, valor in dicionario.items():
    print(valor)

1
2


## Formas de obter as COLUNAS de um df

### Com o tipo Index

In [ ]:
df.columns

Index(['id', 'default', 'idade', 'sexo', 'dependentes', 'escolaridade',
       'estado_civil', 'salario_anual', 'tipo_cartao',
       'meses_de_relacionamento', 'qtd_produtos', 'iteracoes_12m',
       'meses_inativo_12m', 'limite_credito', 'valor_transacoes_12m',
       'qtd_transacoes_12m'],
      dtype='object')

### Transformar em uma lista

In [ ]:
list(df.columns)

['id',
 'default',
 'idade',
 'sexo',
 'dependentes',
 'escolaridade',
 'estado_civil',
 'salario_anual',
 'tipo_cartao',
 'meses_de_relacionamento',
 'qtd_produtos',
 'iteracoes_12m',
 'meses_inativo_12m',
 'limite_credito',
 'valor_transacoes_12m',
 'qtd_transacoes_12m']

### Com .to_list (mais elegante)

In [ ]:
df.columns.to_list()

['id',
 'default',
 'idade',
 'sexo',
 'dependentes',
 'escolaridade',
 'estado_civil',
 'salario_anual',
 'tipo_cartao',
 'meses_de_relacionamento',
 'qtd_produtos',
 'iteracoes_12m',
 'meses_inativo_12m',
 'limite_credito',
 'valor_transacoes_12m',
 'qtd_transacoes_12m']

### Com print de uma coluna por linha

In [ ]:
for coluna in df.columns:
    print(coluna)

id
default
idade
sexo
dependentes
escolaridade
estado_civil
salario_anual
tipo_cartao
meses_de_relacionamento
qtd_produtos
iteracoes_12m
meses_inativo_12m
limite_credito
valor_transacoes_12m
qtd_transacoes_12m


### Juntar todas em uma string

In [ ]:
print(", ".join(df.columns))

id, default, idade, sexo, dependentes, escolaridade, estado_civil, salario_anual, tipo_cartao, meses_de_relacionamento, qtd_produtos, iteracoes_12m, meses_inativo_12m, limite_credito, valor_transacoes_12m, qtd_transacoes_12m
